src/nsrt_learning/side_predicate_learning.py

```diff
@@ -280,3 +280,180 @@ def _get_successor_with_correct_option(
+                                               for a in pnad.op.add_effects}
+                # Update the current image based on the PNAD preconditions.
+                pnad_ground_preconditions = {a.ground(var_to_obj)
+                                             for a in pnad.op.preconditions}
+                current_image -= pnad_ground_add_effects
+                current_image |= pnad_ground_preconditions
+        return current_pnads
+
+
+    def _induce_pnads_from_spec_and_effects(self,
+        option_spec_effects_dict: Dict[ParameterizedOption, Dict[str, Any]]
+        ) -> List[PartialNSRTAndDatastore]:
+        # This is quite hacky and should be refactored.
+        # First, set up PNADs where all of the preconditions are True.
+        pnads = []
+        for param_option, spec_and_effects in option_spec_effects_dict.items():
+            name = f"{param_option.name}LearnedOp"
+            parameter_set = set(spec_and_effects["option_vars"])
+            for atom in spec_and_effects["add"] | spec_and_effects["delete"]:
+                parameter_set.update(atom.variables)
+            parameters = sorted(parameter_set)
+            op = STRIPSOperator(
+                name,
+                parameters,
+                set(),  # preconditions always True
+                spec_and_effects["add"],
+                spec_and_effects["delete"],
+                spec_and_effects["side"])
+            option_spec = (param_option, spec_and_effects["option_vars"])
+            datastore = []  # will get filled in later
+            pnad = PartialNSRTAndDatastore(op, datastore, option_spec)
+            pnads.append(pnad)
+        # Sort the data into the PNADs.
+        self._recompute_datastores_from_segments(pnads)
+        # Update the preconditions based on the sorted data.
+        # Also: remove any side predicates that are no longer needed.
+        # TODO: this is copy and pasted fom strips_learning.py. Refactor.
+        for pnad in pnads:
+            new_side_predicates = set()
+            for i, (segment, var_to_obj) in enumerate(pnad.datastore):
+                objects = set(var_to_obj.values())
+                obj_to_var = {o: v for v, o in var_to_obj.items()}
+                atoms = {
+                    atom
+                    for atom in segment.init_atoms
+                    if all(o in objects for o in atom.objects)
+                }
+                lifted_atoms = {atom.lift(obj_to_var) for atom in atoms}
+                if i == 0:
+                    variables = sorted(set(var_to_obj.keys()))
+                else:
+                    assert variables == sorted(set(var_to_obj.keys()))
+                if i == 0:
+                    preconditions = lifted_atoms
+                else:
+                    preconditions &= lifted_atoms
+                assert len(set(var_to_obj.values())) == len(var_to_obj)
+                # Update the side predicates.
```

Comment:
could you explain the intuition of what's going on here? i don't understand why it fixes the Grasped issue ([722b0e2a3ff4b68a53945277e1a569345d537a4b](https://github.com/Learning-and-Intelligent-Systems/predicators/pull/604/commits/722b0e2a3ff4b68a53945277e1a569345d537a4b))

Reply:
Grasped would only be included as a side predicate now if there were some add effect that included an "unmodeled" Grasped add effect. since there's only ever one Grasped add effect in the segments, and it's modeled in the operator's add effects, Grasped is no longer a side predicate.